source/slang/slang-check-decl.cpp

```diff
@@ -8475,19 +8829,327 @@ namespace Slang
+
+                        continue;
+                    }
+                }
+
+                // Part 2
+                // Regular assignment logic
+                paramIndex++;
+                auto paramType = param->getType();
+                auto paramExpr = m_astBuilder->create<VarExpr>();
+                paramExpr->scope = ctor->ownedScope;
+                paramExpr->declRef = param;
+                paramExpr->type = paramType;
+                paramExpr->loc = param->loc;
+
+                // Do not map a variable which is not visible
+                // Do not map a read-only variable for default-init
+                // Do not map compiler generated 'internal processing' variable
+                VarDeclBase* member = members[memberIndex++].getDecl();
+                while (getDeclVisibility(member) < ctorVisibility
+                    || !getTypeForDeclRef(m_astBuilder, member, member->loc).isLeftValue
+                    || member->getName() && member->getName()->text[0] == '$'
+                    )
+                {
+                    // Should note be possible to be out of range unless compiler generated a synth-ctor wrong.
+                    SLANG_ASSERT(memberIndex < members.getCount());
+                    member = members[memberIndex++].getDecl();
+                }
+
+                // We need to ensure member is `no_diff` if it cannot be differentiated, `ctor` modifiers do not matter
+                // in this case since member-wise ctor is always differentiable or "treat as differentiable".
+                if (!isTypeDifferentiable(member->getType()) || member->hasModifier<NoDiffModifier>())
+                {
+                    auto noDiffMod = m_astBuilder->create<NoDiffModifier>();
+                    noDiffMod->loc = param->loc;
+                    addModifier(param, noDiffMod);
+                }
+
+                addCudaHostModifierIfRequired(ctor, member, foundCudaHostModifier);
+
+                MemberExpr* memberExpr = m_astBuilder->create<MemberExpr>();
+                memberExpr->baseExpression = thisExpr;
+                memberExpr->declRef = member->getDefaultDeclRef();
+                memberExpr->scope = ctor->ownedScope;
+                memberExpr->loc = member->loc;
+                memberExpr->name = member->getName();
+                memberExpr->type = DeclRefType::create(getASTBuilder(), memberExpr->declRef);
+
+                auto assign = m_astBuilder->create<AssignExpr>();
+                assign->left = memberExpr;
+                assign->right = paramExpr;
+                assign->loc = paramExpr->loc;
+
+                auto stmt = m_astBuilder->create<ExpressionStmt>();
+                stmt->expression = assign;
+                stmt->loc = assign->loc;
+
+                seqStmtChild->stmts.add(stmt);
+            }
+            seqStmt->stmts.insert(ctorInfo.m_insertOffset++, seqStmtChild);
+        }
+
+        // Compiler generated ctor may be destroyed
```

Comment:
What will go wrong if we just leave the empty ctor there without deleting it?

Reply:
I just figure out why Ariel wants to delete the empty default ctor.
Considering this case
```
struct S
{
    int a;
}

S s;
```
In this case, based on currently logic, we will synthesize a empty default ctor: `S.$init`, and for `S s;`, we will call this default ctor `S.$init`. So this will cause an issue that slang will report a warning that `a` is not initialized.
While, if there is no default ctor (removing the default ctor), the default ctor won't be called to construct `s`, no warning will be reported.

I didn't participant to your previous discussion about what the default behavior it is. I know that for C++, the default ctor will be called, and C++ won't report warning for that.

If we still want to insist this behavior, I think the warning reported here is quite helppful because it kind of reminds developers to initialize it or use `s = {}` to zero init it.

BTW, there is actually noting wrong that if we don't delete this ctor.

Reply:
@csyonghe to comment this reply first.

Reply:
It is worth noting that there was a second reason that makes this change slightly problematic.

By auto generating `init()` functions, we will be changing how code-gen is structured since certain previously uninitialized `struct` values will now have an assignment to a function that effectively does nothing.

This means that:
1. Certain tests will need to be adjusted.
2. Code-gen will include `init()` functions that do nothing but are still used in code-gen.

Reply:
create sub-task to remove this hack: Create a sub-task to remove this hack: https://github.com/shader-slang/slang/issues/5138